Cleanup execution of external programs. Fix #768

MeanSquaredError opened the pull request:
This is the PR following the discussion in #768 and after implementing what I suggested in #802 

Basically TXS has many places where it invokes different external programs. Two of the most often used ones are in ProcessX (where it executes txs:/// commands) and LatexStyleParser (where it invokes kpathsea/miktex external programs in order to parse tex files and generate .cwl definitions).

In all these places the code uses different ad-hoc implementations to 
1. Adjust the system PATH so that QProcess would be able to find executables without absolute path.
2. Wait for program to complete.
3. Check whether it completed successfully.
4. Read its output.

All these adhoc solutions use different approaches, #ifdefs for the different platforms and generally some of them may work for one platform but others will fail, just like with #768 where running of the build chain from ProcessX worked but running pdflatex from Latexstyleparser failed. Additional problem was that different solutions use different mixes of latin1, utf8 and local text encoding which is another source of possible issues (the correct approach as implied by qgetenv/qputenv is to use only local 8-bit encoding).

So what this patch does is to 

1. Implement a simple class ExecProgram which has two wrapper methods around QProcess.start().

ExecProgram.execAndWait (synchronous)
ExecProgram.execAndNoWait (asynchronous)

Both wrappers adjust the path to include the paths visible from the command shell and optionally to include caller-provided additional search directories.

2. Change LatexStyleParser, ProcessX, KpathseaParser and a couple of other places to use ExecProgram instead of ExecCommand().

3. Remove the no-longer used ExecCommand

sunderme commented on src/latexstyleparser.cpp:
```diff
@@ -736,28 +734,18 @@ QStringList LatexStyleParser::readPackageTexDef(QString fn) const
-	myProc.waitForFinished();
-	if (myProc.exitCode() != 0)
+	execProgram.execAndWait ();
+	if (!execProgram.normalRun || (execProgram.exitCode != 0)) {
```
to my taste, a method like runFinishedWithoutError could simply things here (and later)

MeanSquaredError replied:
OK, changed it so that `execAndWait` effectively returns 
`normalRun && (execProgram.exitCode == 0)`
So now we are checking the result when calling `execProgram.execAndWait`

sunderme commented on src/latexstyleparser.cpp:
```diff
@@ -823,29 +811,19 @@ QStringList LatexStyleParser::readPackageTracing(QString fn) const
+	execProgram.arguments << "-draftmode" << "-interaction=nonstopmode" << "--disable-installer" << tf->fileName();
+	execProgram.workingDirectory = QFileInfo(tf->fileName()).absoluteDir().absolutePath();
+	execProgram.execAndWait();
+	if (!execProgram.normalRun || (execProgram.exitCode != 0)) {
```
like here, see above

MeanSquaredError replied:
OK, changed it so that `execAndWait` effectively returns 
`normalRun && (execProgram.exitCode == 0)`
So now we are checking the result when calling `execProgram.execAndWait`

sunderme commented on src/execprogram.h:
```diff
@@ -0,0 +1,25 @@
+class ExecProgram
+{
+public:
+	// Input parameters
```
personally, I would like an constructor which allows setting the main parameters ...

MeanSquaredError replied:
Added the constructors.

MeanSquaredError commented:
OK, I changed the execAndWait so that it returns a boolean indicating normal termination with exit code 0. Also simplified the code that call execAndWait to check the return value.
Did not rename execAndWait -> runFinishedWithoutError, because I could not come up with a good new name for execAndNoWait, but if you have a good name for it too, then I can rename both methods.

Now what is left is to add the constructor(s).

MeanSquaredError commented:
OK, I added three different constructors to ExecProgram to initialize the member variables and changed the calling code to use the constructors.

I also prefixed the names of the data members of ExecProgram with m_ to prevent name clashes of the data members with the constructor arguments.

I think the PR can now undergo another review.

MeanSquaredError commented:
Ooops, Codacy complained :-) let me fix it.

MeanSquaredError commented:
OK, fixed the issues by initializing m_normalRun and moving all the assignments to the constructor initializer list. Hopefully Codacy is OK with it now.

sunderme commented:
looks good

sunderme commented:
thanks

muzimuzhi commented:
@MeanSquaredError Many thanks.
